### PR TITLE
Refactor yum and dnf, add feature parity

### PR DESCRIPTION
**Pull request by maxamillion**
Signed-off-by: Adam Miller <admiller@redhat.com>

##### SUMMARY
Bringing the `yum` and `dnf` modules into feature parity and move both to an object oriented model to reduce redundant information passing between functions (hopefully making them both more maintainable). 

A main goal of this is for the `dnf` module to have feature parity with the `yum` module but also be able to extend beyond that of the `yum` module as upstream `dnf` continues to add more features, this is in service of allowing the `yum` Ansible module to remain compatible with [YUM4](https://wiki.centos.org/SpecialInterestGroup/ConfigManagementSIG/YUM4) which actually has a dnf backend. (The compatibility will be handled with a yum action plugin in a separate pull request, I wanted to try and limit the scope of each PR)

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

yum
dnf

##### ANSIBLE VERSION

```
ansible 2.7.0.dev0 (refactor/yum_dnf_parity 29ecf9b5b7) last updated 2018/08/02 16:44:51 (GMT -500)
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/home/admiller/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/admiller/src/dev/ansible/lib/ansible
  executable location = /home/admiller/.virtualenvs/ansible/bin/ansible
  python version = 2.7.15 (default, May 16 2018, 17:50:09) [GCC 8.1.1 20180502 (Red Hat 8.1.1-1)]
```

##### ADDITIONAL INFORMATION

```

```

### CODE REVIEW

File: lib/ansible/module_utils/yumdnf.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+        if self.disablerepo and len(self.disablerepo) == 1 and ',' in self.disablerepo:
+            self.disablerepo = self.module.params['disablerepo'].split(',')
+        if self.exclude and len(self.exclude) == 1 and ',' in self.exclude:
+            self.exclude = self.module.params['exclude'].split(',')
+
+    @abstractmethod
+    def run(self):
+        raise NotImplementedError
+
+    def fetch_rpm_from_url(self, spec):
```
Comment (mkrizek) [inline lib/ansible/module_utils/yumdnf.py]: Hopefully we'll move this into utils soon https://github.com/ansible/ansible/pull/19172

File: lib/ansible/module_utils/yumdnf.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+# 64k.  Number of bytes to read at a time when manually downloading pkgs via a url
+BUFSIZE = 65536
+
+
+class YumDnf(with_metaclass(ABCMeta, object)):
+    """
+    Abstract class that handles the population of instance variables that should
+    be identical between both YUM and DNF modules because of the feature parity
+    and shared argement spec
```
Comment (mkrizek) [inline lib/ansible/module_utils/yumdnf.py]: s/argement/argument :)

File: test/integration/targets/dnf/tasks/dnf.yml [hunk 3]
```diff
@@ -241,6 +274,8 @@
     state: present
   register: dnf_result
 
+- debug: var=dnf_result
```
Comment (mkrizek) [inline test/integration/targets/dnf/tasks/dnf.yml]: I wonder if we should put something like `verbosity: 1` for debug prints?

File: lib/ansible/modules/packaging/os/yum.py [hunk 4]
```diff
@@ -146,6 +150,22 @@
         The disabled plugins will not persist beyond the transaction.
     required: false
     version_added: "2.5"
+  releasever:
```
Comment (mkrizek) [inline lib/ansible/modules/packaging/os/yum.py:153]: It makes me a little nervous that we're not adding tests for new features.

**Review by mkrizek — COMMENTED** [review]
I love that we're getting rid of the long method signatures by extracting the arguments into instance vars +1000. It's huge diff to review, since all tests pass I think we should merge ASAP if we want this in `2.7` so we can have that tested properly in `devel`.

File: lib/ansible/modules/packaging/os/yum.py (hunk 4 above)
Comment (maxamillion) [reply lib/ansible/modules/packaging/os/yum.py:153]: Yeah, I should really add more tests. The problem is that I'm not sure how best to test this just yet. Let me think about it a little and try to get something added that has a reasonable assertion.

Comment (maxamillion) [reply lib/ansible/modules/packaging/os/yum.py:153]: However, we don't have test coverage of that functionality for the `yum` module and I brought over roughly a dozen new options to `dnf` from `yum`.... but I do agree we still need test coverage. I'll look at this when I'm back.

File: test/integration/targets/dnf/tasks/dnf.yml (hunk 3 above)
Comment (maxamillion) [reply test/integration/targets/dnf/tasks/dnf.yml]: @mkrizek that was actually an accidental left-over for when I was working on this, it's not needed for the tests

**Comment by maxamillion** [comment]
As per @abadger on irc, as 2.7 release manager, this module refactor was approved (mentioned here for posterity)

**Comment by maxamillion** [comment]
rebuild_merge

**Comment by sej7278** [comment]
This doesn't seem to work for me, at least not in the same way as ```package-cleanup --leaves``` or ```yum autoremove``` do.

plays tried:

```
- name: prune packages
  yum: name=* autoremove=yes

- name: remove some packages
  yum:
    name:
      - kernel-headers
      - blah
    state: absent
    autoremove: yes
```

doesn't remove anything other than what i've asked it to remove, but then if i run ```yum autoremove``` it suggests this lot:

```
# yum autoremove
Loaded plugins: ulninfo
Resolving Dependencies
--> Running transaction check
---> Package fxload.x86_64 0:2002_04_11-16.el7 will be erased
---> Package libndp.x86_64 0:1.2-7.el7 will be erased
---> Package lsscsi.x86_64 0:0.27-6.el7 will be erased
---> Package mariadb-libs.x86_64 1:5.5.60-1.el7_5 will be erased
---> Package teamd.x86_64 0:1.27-5.el7 will be erased
--> Finished Dependency Resolution
--> Finding unneeded leftover dependencies
---> Marking libdaemon to be removed - no longer needed by teamd
---> Marking libnl3 to be removed - no longer needed by teamd
---> Marking libteam to be removed - no longer needed by teamd
---> Marking libnl3-cli to be removed - no longer needed by teamd
Found and removing 4 unneeded dependencies
--> Running transaction check
---> Package libdaemon.x86_64 0:0.14-7.el7 will be erased
---> Package libnl3.x86_64 0:3.2.28-4.el7 will be erased
---> Package libnl3-cli.x86_64 0:3.2.28-4.el7 will be erased
---> Package libteam.x86_64 0:1.27-5.el7 will be erased
--> Finished Dependency Resolution
```

ansible 2.7.1 and oel 7.6 (yum 3.4.3, rpm 4.11)

**Comment by ghost** [comment]
I have the same issue as @sej7278 !

**Comment by ghost** [comment]
Seems that according to the actual PR (https://github.com/ansible/ansible/pull/28700/), this is not possible.

I resorted to:

```
- name: Clean unused packages
  command: yum autoremove -y warn=no
```

**Comment by sej7278** [comment]
@alex-mohemian yes that's what i've been using but not too happy with shelling out (and hiding the warning about doing so!) can't find where it says its not possible though.

i'm also not happy about using the term "autoremove" in the yum module if its clearly not doing what ```yum autoremove``` does; the docs aren't at all clear about it either - something about "user installed" packages only being catered for.

**Comment by maxamillion** [comment]
@alex-mohemian @sej7278  Yes, `autoremove` was broken by mistake. The issue should be fixed once Ansible 2.7.2 is released https://github.com/ansible/ansible/pull/47902

**Comment by ghost** [comment]
@maxamillion thank you. Are you sure the behavior is fixed and that 

```
- name: cleanup
  yum: autoremove=yes
```

will do what `sudo yum autoremove -y` does? From https://github.com/ansible/ansible/pull/47902 I don't see that...

**Comment by sej7278** [comment]
just tried rhel8 beta1 and ansible is totally broken (not least by no python interpreters!)

but these don't work at all:

```
- name: prune packages
  dnf:
    autoremove: yes

- name: prune packages
  dnf: autoremove=yes

fatal: [vbrhel8]: FAILED! => {"changed": false, "msg": "Autoremove should be used alone or with state=absent", "results": []}

```

**Comment by maxamillion** [comment]
@sej7278 did you try with current devel or wtih the `stable-2.7` branch from source? the fix is still unreleased

**Comment by maxamillion** [comment]
@alex-mohemian yes, the issue was that the value was being set incorrectly for `self.autoremove` and `self.state` that were being checked to then invoke `yum -y autoremove` so the change only sets those correctly, the conditional to set the command call on the backend didn't need modification.

**Comment by sej7278** [comment]
i'm using ansible debian package 2.7.1+dfsg-2 which i assume has neither yum nor dnf autoremove fixes (its from about 28th october)

oddly enough even regular removal doesn't work (on rhel8b1) for example:

```
- name: remove packages
  dnf:
    name:
      - kernel-headers
      - iwl*firmware*
      - NetworkManager*
    state: absent
    autoremove: yes
```

or using yum.

**edit:** seems dnf doesn't like wildcards anymore, but yum used to work....? https://github.com/ansible/ansible/pull/45357

**Comment by sej7278** [comment]
ok this is quite odd, if i remove the wildcards which we know is broken (until 2.8.0?) then autoremove works - this removes teamd, libdaemon etc; the same as ```dnf -y autoremove``` would:

```
- name: remove packages
  dnf:
    name:
      - NetworkManager-team
      - NetworkManager-libnm
      - NetworkManager-tui
      - NetworkManager
    state: absent
    autoremove: yes
```

but autoremove on its own doesn't work, and this same play works using yum too on rhe8b1 so maybe its rpm3.4.3 from rhel7 that's the real problem?

**Pull request merged**